api/pipeline.go

```diff
@@ -0,0 +1,647 @@
+
+		return
+	}
+
+	// send API call to capture the pipeline configuration file
+	config, err := source.FromContext(c).ConfigBackoff(u, r.GetOrg(), r.GetName(), ref)
+	if err != nil {
+		retErr := fmt.Errorf("unable to get pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
+
+		util.HandleError(c, http.StatusNotFound, retErr)
+
+		return
+	}
+
+	// create the compiler with extra information embedded into it
+	comp := compiler.FromContext(c).
+		WithMetadata(m).
+		WithRepo(r).
+		WithUser(u)
+
+	// parse the pipeline configuration file
+	p, err := comp.Parse(config)
+	if err != nil {
+		retErr := fmt.Errorf("unable to parse pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	// create map of templates for easy lookup
+	t := mapFromTemplates(p.Templates)
+
+	// check if the pipeline contains stages
+	if len(p.Stages) > 0 {
+		// inject the templates into the stages
+		p.Stages, err = comp.ExpandStages(p.Stages, t)
+		if err != nil {
+			retErr := fmt.Errorf("unable to expand stages in pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
+
+			util.HandleError(c, http.StatusBadRequest, retErr)
+
+			return
+		}
+
+		// inject the substituted environment variables into the stages
+		p.Stages, err = comp.SubstituteStages(p.Stages)
+		if err != nil {
+			retErr := fmt.Errorf("unable to substitute stages in pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
+
+			util.HandleError(c, http.StatusBadRequest, retErr)
+
+			return
+		}
+
+	} else {
+		// inject the templates into the steps
+		p.Steps, err = comp.ExpandSteps(p.Steps, t)
+		if err != nil {
+			retErr := fmt.Errorf("unable to expand steps in pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
+
+			util.HandleError(c, http.StatusBadRequest, retErr)
+
+			return
+		}
+
+		// inject the substituted environment variables into the steps
+		p.Steps, err = comp.SubstituteSteps(p.Steps)
+		if err != nil {
+			retErr := fmt.Errorf("unable to substitute steps in pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
+
+			util.HandleError(c, http.StatusBadRequest, retErr)
+
+			return
+		}
+
+	}
+
+	// validate the yaml configuration
+	err = comp.Validate(p)
+	if err != nil {
+		retErr := fmt.Errorf("unable to validate pipeline configuration for %s@%s: %w", r.GetFullName(), ref, err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	// format response body based off output query parameter
+	switch strings.ToLower(output) {
+	case outputJSON:
+		c.JSON(http.StatusOK, p)
+	case outputYAML:
+		fallthrough
+	default:
+		c.YAML(http.StatusOK, p)
+	}
+}
+
+// helper function that creates a map of templates from a yaml configuration.
+func mapFromTemplates(templates []*yaml.Template) map[string]*yaml.Template {
+	m := make(map[string]*yaml.Template)
+
+	for _, tmpl := range templates {
+		m[tmpl.Name] = tmpl
+	}
+
+	return m
+}
```

Comment:
Do you think it's worth considering to create this map object in `types` for templates?

We have an identical function in the [go-vela/compiler](https://github.com/go-vela/compiler) repo 👍 

https://github.com/go-vela/compiler/blob/master/compiler/native/expand.go#L92-L101

Reply:
Yeah I think we discussed with that Id also like to add a type for Template that has a Link field for use with the UI.

Reply:
It was going to be roped into an enhancement PR for /templates but I could just do it in this PR

Reply:
If you plan to do it in a different PR then I think that is totally fine 👍

Reply:
I'll just do it now, rather not merge in the duplicate code and risk forgetting about it 😅

Reply:
done 👍